src/language-services/analysis.ts

```diff
@@ -15,53 +19,50 @@ import {
     SignatureProviderContext,
     SymbolProvider,
 } from "./providers";
+import * as WorkspaceCache from "./workspaceCache";
 
-export interface Analysis {
+export interface Analysis extends IDisposable {
     getCompletionItems(): Promise<CompletionItem[]>;
     getHover(): Promise<Hover>;
     getSignatureHelp(): Promise<SignatureHelp>;
 }
 
-export interface AnalysisOptions {
-    readonly environmentSymbolProvider?: SymbolProvider;
-    readonly librarySymbolProvider?: LibrarySymbolProvider;
-}
-
-export function createAnalysisSession(document: TextDocument, position: Position, options: AnalysisOptions): Analysis {
+export function createAnalysisSession(document: TextDocument, position: Position, options: Configuration): Analysis {
     return new DocumentAnalysis(document, position, options);
 }
 
-class DocumentAnalysis implements Analysis {
-    private readonly maybeTriedInspection: PQP.Task.TriedInspection | undefined;
-    private readonly environmentSymbolProvider: SymbolProvider;
-    private readonly keywordProvider: KeywordProvider;
-    private readonly librarySymbolProvider: LibrarySymbolProvider;
-    private readonly localSymbolProvider: SymbolProvider;
-
-    constructor(
-        private readonly document: TextDocument,
-        private readonly position: Position,
-        options: AnalysisOptions,
-    ) {
-        this.maybeTriedInspection = WorkspaceCache.maybeTriedInspection(this.document, this.position);
-
-        this.environmentSymbolProvider = options.environmentSymbolProvider
-            ? options.environmentSymbolProvider
+abstract class AnalysisBase implements Analysis {
+    protected readonly environmentSymbolProvider: SymbolProvider;
+    protected readonly keywordProvider: KeywordProvider;
+    protected readonly librarySymbolProvider: LibrarySymbolProvider;
+    protected readonly localSymbolProvider: SymbolProvider;
+
+    protected readonly options: Configuration;
+    protected readonly position: Position;
+    protected readonly triedInspection: PQP.Task.TriedInspection | undefined;
+
+    constructor(triedInspection: PQP.Task.TriedInspection | undefined, position: Position, options: Configuration) {
+        this.triedInspection = triedInspection;
+        this.options = options;
```

Comment:
Do we want to store options as a member? I think we only need to store `maintainWorkspaceCache` as the rest are used only in the constructor.

Reply:
I don't see a problem with storing it, and I thought the subclasses might want to be able to access it. I'll revisit this when I do the refactoring around the use of the workspace cache - I might end up going down to a single implementation anyways.

Reply:
Sounds good to me.